Cluster Transport Encryption Documentation

DSchalla opened the pull request:
This pull request contains a draft for the explanation how to set up a Mattermost cluster with internode cluster encryption using SSH tunneling.

amyblais commented on source/install/config-cluster-transport-encryption.rst:
```diff
@@ -0,0 +1,220 @@
+by no means limited to that amount.
+
+**Prerequisites:**
+- SSH port whitelisted between each node of the deployment
```
These are not rendering as a list.

amyblais commented on source/install/config-cluster-transport-encryption.rst:
```diff
@@ -0,0 +1,220 @@
+=============================================
+Configuring Cluster Transport Encryption (E20)
```
This can also be added to the table of contents.

amyblais commented on source/install/config-cluster-transport-encryption.rst:
```diff
@@ -0,0 +1,220 @@
+
+- Connect to each Mattermost server with a sudo or root user
+- Make a note of the IP from each cluster member used for the internal communication
+- Ensure `AllowTcpForwarding` is enabled in `/etc/ssh/sshd_config` of each cluster node
```
```suggestion
- Ensure ``AllowTcpForwarding`` is enabled in ``/etc/ssh/sshd_config`` of each cluster node
```

amyblais commented on source/install/config-cluster-transport-encryption.rst:
```diff
@@ -0,0 +1,220 @@
+the usage of another storage location.
+
+Next, ensure that the SSH public key of each node is added to `authorized_keys`
+file of the other nodes of the cluster. To do so, copy the contents of `/home/mattermost/.ssh/id_rsa.pub`
```
```suggestion
file of the other nodes of the cluster. To do so, copy the contents of ``/home/mattermost/.ssh/id_rsa.pub``
```

amyblais commented on source/install/config-cluster-transport-encryption.rst:
```diff
@@ -0,0 +1,220 @@
+
+Next, ensure that the SSH public key of each node is added to `authorized_keys`
+file of the other nodes of the cluster. To do so, copy the contents of `/home/mattermost/.ssh/id_rsa.pub`
+of node 2 and 3, and add it to `/home/mattermost/.ssh/authorized_keys` of node 1.
```
```suggestion
of node 2 and 3, and add it to ``/home/mattermost/.ssh/authorized_keys`` of node 1.
```

amyblais commented on source/install/config-cluster-transport-encryption.rst:
```diff
@@ -0,0 +1,220 @@
+The location of the SSH key itself is irrelevant if company policies require
+the usage of another storage location.
+
+Next, ensure that the SSH public key of each node is added to `authorized_keys`
```
```suggestion
Next, ensure that the SSH public key of each node is added to ``authorized_keys``
```

amyblais commented on source/install/config-cluster-transport-encryption.rst:
````diff
@@ -0,0 +1,220 @@
+```
+
+Repeat the same steps on the other nodes, but replacing the IPs with the ones from the
+other member nodes. Do so for each membernode, excluding the node itself.
````
```suggestion
other member nodes. Do so for each member node, excluding the node itself.
```

amyblais commented on source/install/config-cluster-transport-encryption.rst:
```diff
@@ -0,0 +1,220 @@
+Repeat the same steps on the other nodes, but replacing the IPs with the ones from the
+other member nodes. Do so for each membernode, excluding the node itself.
+
+As a next step, open `/etc/ufw/after.rules` and add the following block to the
```
```suggestion
As a next step, open ``/etc/ufw/after.rules`` and add the following block to the
```

amyblais commented on source/install/config-cluster-transport-encryption.rst:
````diff
@@ -0,0 +1,220 @@
+-A OUTPUT -p tcp -d ip_node_4 --dport 8074 -j DNAT --to-destination 127.0.0.1:38074
+```
+
+Please be aware that the ports on the right side must be unique, so if having a cluster of
````
```suggestion
Please be aware that the ports on the right side must be unique, so if you have a cluster of
```

amyblais commented on source/install/config-cluster-transport-encryption.rst:
````diff
@@ -0,0 +1,220 @@
+```
+
+Please be aware that the ports on the right side must be unique, so if having a cluster of
+6 nodes use 8075 and 8074 with 1 to 5 in front of it. If the cluster is of bigger size additional
````
```suggestion
6 nodes, use 8075 and 8074 with 1 to 5 in front of it. If the cluster is of bigger size additional
```

amyblais commented on source/install/config-cluster-transport-encryption.rst:
```diff
@@ -0,0 +1,220 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As a next step, we will ensure that the SSH tunnels are created as part of the Mattermost service
+start. To do so, create a file called `pre_start.sh` in `/opt/mattermost/bin` on `mattermost1`:
```
```suggestion
start. To do so, create a file called ``pre_start.sh`` in ``/opt/mattermost/bin`` on ``mattermost1``:
```

amyblais commented on source/install/config-cluster-transport-encryption.rst:
````diff
@@ -0,0 +1,220 @@
+$ chmod +x /opt/mattermost/bin/pre_start.sh
+```
+
+Open the systemd unit file of Mattermost and search for `Type=Notify`, after it enter
````
```suggestion
Open the systemd unit file of Mattermost and search for ``Type=Notify``, after it enter
```

amyblais commented on source/install/config-cluster-transport-encryption.rst:
````diff
@@ -0,0 +1,220 @@
+```
+
+Open the systemd unit file of Mattermost and search for `Type=Notify`, after it enter
+a `ExecStartPre` script that will be executed before Mattermost itself is started:
````
```suggestion
a ``ExecStartPre`` script that will be executed before Mattermost itself is started:
```

amyblais commented on source/install/config-cluster-transport-encryption.rst:
````diff
@@ -0,0 +1,220 @@
+On each node generate a SSH key-pair for the service account, in our scenario
+called `mattermost`:
+
+```
````
Please see https://docs.mattermost.com/process/documentation-guidelines.html#literal-blocks for how to add code blocks.

DSchalla commented:
Thanks a ton for the in-depth review, and fixing the code blocks @amyblais! I was still in automatic mode with markdown.

grundleborg commented on source/install/config-cluster-transport-encryption.rst:
````diff
@@ -0,0 +1,220 @@
+follows:
+```
+$ sudo ufw allow from 10.10.250.231/32 to any port ssh
+Skipping adding existing rule
````
huh?

DSchalla replied:
Yup, will fix. I already had that executed, so the output was a bit weird. :-)

DSchalla commented:
I've fixed most of the formatting bugs, added an overview page and a database transport encryption page. I'll work Monday last on a page about the encryption from Proxy to Mattermost and including it in the ToC, then we are good to go.

mattermod commented:
This issue has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @hanzei

jasonblais commented:
/cc @DSchalla (though I'm sure this is in your radar to revisit)

DSchalla commented:
@jasonblais Yeah, as marked by "WIP", it is indeed still in progress. Last review from myself later today and then ready for PM/Dev review.

DSchalla commented:
Seems Atom removed some unnecessary whitespace in the process in the `source/guides/administrator.rst`, since that won't be any impact I think reverting those changes isn't required?

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+using SSH tunneling. The guide walks through the deployment of this solution on
+Ubuntu 18.04, but it can be adapted for any Linux operating system.
+
+While this document only describes the configuration of a 3 node cluster, it is
```
```suggestion
While this document only describes the configuration of a three-node cluster, it is
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+While this document only describes the configuration of a 3 node cluster, it is
+by no means limited to that amount.
+
+**Prerequisites:**
```
```suggestion
## Prerequisites
```

DSchalla commented:
Thanks @justinegeffen, updated. :+1:

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+Mattermost is able to encrypt the traffic between the proxy and the application
+server using TLS.
+
+**Prerequisites:**
```
```suggestion
## Prerequisites
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+using TLS. This guide will walk you through the set up steps for a single, separate MySQL
+server.
+
+**Prerequisites:**
```
```suggestion
## Prerequisites
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+- Mattermost running with a dedicated service user
+- Mattermost service is stopped on each cluster node
+
+**Note:** A support on the application level is currently in development and will
```
```suggestion
**Note:** Support on the application level is currently in development and will
```
Not sure if that change works - I may have misunderstood the sentence. From what I can see, it refers to support offered by us on the application level?

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+by no means limited to that amount.
+
+## Prerequisites
+- SSH port whitelisted between each node of the deployment
```
Do we need to provide a port number? Is this done on Mattermost?

DSchalla replied:
No, this is usually a standard port that may depend on the persons configuration as well. We can safely assume they know this.

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+
+## Prerequisites
+- SSH port whitelisted between each node of the deployment
+- ufw / iptables is active on each node
```
```suggestion
- Active ufw/iptables on each node
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+## Prerequisites
+- SSH port whitelisted between each node of the deployment
+- ufw / iptables is active on each node
+- Access to the root / sudo user of each node for configuration
```
```suggestion
- Access to the root/sudo user of each node for configuration
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+Example Environment
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+In this scenario there are 3 application nodes in our environment with the following
+hostname / IP mapping:
```
```suggestion
hostname/IP mapping:
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+
+Example Environment
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+In this scenario there are 3 application nodes in our environment with the following
```
```suggestion
In this scenario there are three application nodes in our environment with the following
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+The location of the SSH key itself is irrelevant if company policies require
+the usage of another storage location.
+
+Next, ensure that the SSH public key of each node is added to ``authorized_keys``
```
```suggestion
Next, ensure that the SSH public key of each node is added to the ``authorized_keys``
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+
+Next, ensure that the SSH public key of each node is added to ``authorized_keys``
+file of the other nodes of the cluster. To do so, copy the contents of ``/home/mattermost/.ssh/id_rsa.pub``
+of node 2 and 3, and add it to ``/home/mattermost/.ssh/authorized_keys`` of node 1.
```
```suggestion
of nodes 2 and 3, and add it to ``/home/mattermost/.ssh/authorized_keys`` of node 1.
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+able to establish an SSH connection to the other nodes of the cluster.
+
+**Note:** This service account can be separate from the service account already used
+for the Mattermost systemd service itself. It is only important that it is allowed
```
```suggestion
for the Mattermost ``systemd`` service itself. It is only important that it is allowed
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+  22/tcp                     ALLOW       10.10.250.165
+
+
+Repeat the same steps on the other nodes, but replacing the IPs with the ones from the
```
```suggestion
Repeat the same steps on the other nodes, replacing the IPs with the ones from the
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+Repeat the same steps on the other nodes, but replacing the IPs with the ones from the
+other member nodes. Do so for each member node, excluding the node itself.
+
+As a next step, open ``/etc/ufw/after.rules`` and add the following block to the
```
```suggestion
Next, open ``/etc/ufw/after.rules`` and add the following block to the
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+  COMMIT
+
+
+Two lines always belong to a single node, so in a deployment with 4 nodes:
```
```suggestion
Two lines always belong to a single node, so in a deployment with four nodes:
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+  -A OUTPUT -p tcp -d ip_node_4 --dport 8074 -j DNAT --to-destination 127.0.0.1:38074
+
+Please be aware that the ports on the right side must be unique, so if you have a cluster of
+6 nodes, use 8075 and 8074 with 1 to 5 in front of it. If the cluster is of bigger size, additional
```
```suggestion
six nodes, use 8075 and 8074 with 1 to 5 in front of it. If the cluster is of bigger size, additional
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+
+- SSH access enabled in firewall from each cluster node to another
+- Per node 2 iptables rules for port 8074 and 8075
+- IP Forwarding enabled
```
```suggestion
- IP forwarding enabled
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+Cluster Start
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+After finishing the configuration on each node, start the service on each cluster again
```
```suggestion
Once each node is configured, restart the service on each cluster.
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+After finishing the configuration on each node, start the service on each cluster again
+and confirm it's running:
```
```suggestion
and confirm it's running using the command below.
```

justinegeffen commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+    Process: 16734 ExecStartPre=/opt/mattermost/bin/pre_start.sh (code=exited, status=0/SUCCESS)
+
+
+Afterwards open the Mattermost System Console and confirm that each node is reporting successfully
```
```suggestion
Next, open the Mattermost System Console and confirm that each node is reporting successfully
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+=============================================
+
+Mattermost is able to encrypt the traffic between the database and the application
+using TLS. This guide will walk you through the set up steps for a single, separate MySQL
```
```suggestion
using TLS. This guide describes the set up steps for a single, separate MySQL
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+## Prerequisites
+
+- Mattermost server or cluster operational
+- MySQL server operational
```
```suggestion
- Operational MySQL server 
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+
+## Prerequisites
+
+- Mattermost server or cluster operational
```
```suggestion
- Operational Mattermost server or cluster
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+
+- Mattermost server or cluster operational
+- MySQL server operational
+- Connectivity between Mattermost and MySQL server confirmed
```
```suggestion
- Confirmed connectivity between Mattermost and MySQL server
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+
+Example Environment
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+In this scenario there is 1 Mattermost application server and one MySQL server,
```
```suggestion
In this scenario there is one Mattermost application server and one MySQL server,
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+- **transport-encryption-mattermost1:** 10.10.250.146
+- **transport-encryption-mysql1:** 10.10.250.148
+
+Preparations
```
```suggestion
Preparations
```

I would remove this sub-section and move the instructions to the next sub-section or simply have it as an opening line for the MySQL and Mattermost sub-sections (i.e., start that section with "Connect to the server as sudo/root user").

DSchalla replied:
Sure, that sounds good to me. Let me fix that!

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+
+- Connect to both servers with a sudo or root user
+
+MySQL Configuration
```
```suggestion
Configuring MySQL
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+
+MySQL Configuration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+On the MySQL server, execute the following command to prepare the server for
```
```suggestion
Execute the following command to prepare the server for
```
Removed the first bit because it's in the MySQL section and seems redundant. 3/5.

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+
+``sudo mysql_ssl_rsa_setup --uid=mysql``
+
+This will generate self-signed certificates in ``/var/lib/mysql/`` that the MySQL
```
```suggestion
This generates self-signed certificates in ``/var/lib/mysql/`` that the MySQL
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+``sudo mysql_ssl_rsa_setup --uid=mysql``
+
+This will generate self-signed certificates in ``/var/lib/mysql/`` that the MySQL
+server will use to encrypt the connection. If you would like to use certificates
```
```suggestion
server uses to encrypt the connection. If you would like to use certificates
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+
+This will generate self-signed certificates in ``/var/lib/mysql/`` that the MySQL
+server will use to encrypt the connection. If you would like to use certificates
+from your company CA, please follow the MySQL documentation for configuration steps.
```
```suggestion
from your company CA, please follow the MySQL documentation for configuration steps.
```
Do we have a link to that documentation and is there a need to provide any additional guidance for this in terms of adding the certificate using this option?

DSchalla replied:
Since it depends on the version they are using, we sadly won't be able to provide a deep-link for them here.

justinegeffen replied:
Got it, that makes sense. :)

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+from your company CA, please follow the MySQL documentation for configuration steps.
+
+**Optionally**, it can be enforced that all connections must be made via a local
+socket connection or TLS, to do so, open ``/etc/mysql/mysql.conf.d/mysqld.cnf``
```
```suggestion
socket connection or TLS. To do this, open ``/etc/mysql/mysql.conf.d/mysqld.cnf``
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+
+``require_secure_transport = ON``
+
+Any connection to the MySQL server now must be made with secure transport enabled.
```
```suggestion
Any connection to the MySQL server must now be made with secure transport enabled.
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+    Oct 18 16:41:25 transport-encryption-mysql1 systemd[1]: Started MySQL Community Server.
+
+
+Mattermost Configuration
```
```suggestion
Configuring Mattermost
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+
+If you are running Mattermost in a cluster, be sure to update the value on each node
+of the cluster. If you are using configuration in the database, be sure to update the
+systemd unit file as well to enable TLS for the configuration store.
```
```suggestion
``systemd`` unit file as well to enable TLS for the configuration store.
```

justinegeffen commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,118 @@
+of the cluster. If you are using configuration in the database, be sure to update the
+systemd unit file as well to enable TLS for the configuration store.
+
+Afterwards, restart the Mattermost server and ensure the system is operational:
```
```suggestion
Once complete, restart the Mattermost server and ensure the system is operational:
```

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+
+## Prerequisites
+
+- Mattermost server or cluster operational
```
```suggestion
- Operational Mattermost server or cluster
```

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+
+Example Environment
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+In this scenario there is 1 Mattermost application server and one NGINX server,
```
```suggestion
In this scenario there is one Mattermost application server and one NGINX server,
```

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+Preparations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Connect to both servers with a sudo or root user
```
Same comment as in the previous section.

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+
+  proxy_pass https://backend;
+
+Afterwards do not reload the NGINX server yet to minimize the downtime of the service.
```
```suggestion
To minimize downtime of the service, do not reload the NGINX server yet.
```
Can we provide a next step/suggestion as to when to reload it?

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+
+- Connect to both servers with a sudo or root user
+
+NGINX Configuration
```
```suggestion
Configuring NGINX
```

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+
+Afterwards do not reload the NGINX server yet to minimize the downtime of the service.
+
+Mattermost Configuration
```
```suggestion
Configuring Mattermost
```

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+On the Mattermost server, change to the config directory of Mattermost and generate a self-signed certificate
+that will be used to encrypt the traffic between the proxy server and the application server.
+
+**Note:** As an alternative, sign a certificate from your companies CA.
```
```suggestion
**Note:** Alternatively you can sign a certificate from your company's CA.
```

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+  $ chmod 640 *.pem
+
+
+Afterwards open the file ``config.json`` and modify the values ``ConnectionSecurity``,
```
```suggestion
Once complete, open the file ``config.json`` and modify the values ``ConnectionSecurity``,
```

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+Afterwards open the file ``config.json`` and modify the values ``ConnectionSecurity``,
+``TLSCertFile`` and ``TLSKeyFile`` in the ``ServiceSettings`` section.
+
+**Before:**
```
```suggestion
**Before**
```

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+          "TLSCertFile": "",
+          "TLSKeyFile": "",
+
+**After:**
```
```suggestion
**After**
```

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+      },
+
+
+Afterwards, restart the Mattermost server and ensure it's up and running:
```
```suggestion
Restart the Mattermost server and ensure it's up and running:
```

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+     Active: active (running) since Mon 2019-10-28 16:45:29 UTC; 1h 15min ago
+     [...]
+
+Finally, on the **NGINX server**, reload the configuration that requests are being
```
```suggestion
Finally, on the **NGINX server**, reload the HTTP requests configuration.
```

DSchalla replied:
What you think about `Finally, on the NGINX server, reload the confgiuration to ensure that requests are sent on HTTPS:`?

justinegeffen replied:
Yes, that works well I think! :)

justinegeffen commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,108 @@
+     [...]
+
+Finally, on the **NGINX server**, reload the configuration that requests are being
+made on HTTPS:
```
```suggestion
made on HTTPS:
```

Remove.

justinegeffen commented on source/install/transport-encryption/config.rst:
```diff
@@ -0,0 +1,22 @@
+Configuring Transport Encryption
+=============================================
+
+The components of the Mattermost setup can be shown in the following graphic,
```
```suggestion
The components of the Mattermost setup are shown in the following diagram,
```

justinegeffen commented on source/install/transport-encryption/config.rst:
```diff
@@ -0,0 +1,22 @@
+=============================================
+
+The components of the Mattermost setup can be shown in the following graphic,
+incl. the used transport encryption used. Aside from the encryption between the
```
```suggestion
including the transport encryption used. Aside from the encryption between the
```

DSchalla commented:
@justinegeffen Finally got a change to work on the feedback, thank you very much for the in-depth review! I added some comments and 1-2 questions if you could take a look. :-)

justinegeffen commented:
Thanks, @DSchalla! I like your suggestion and it all looks good to me. :)

DSchalla commented:
@grundleborg Ready for final reviews. :-)

amyblais commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+While this document only describes the configuration of a three-node cluster, it is
+by no means limited to that amount.
+
+## Prerequisites
```
The header formatting should be corrected, and the list formatting is not rendering correctly.

amyblais commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+
+SSH Authentication
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+On each node generate a SSH key-pair for the service account, in our scenario
```
```suggestion
On each node, generate a SSH key-pair for the service account. In our scenario this is
```

amyblais commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+able to establish an SSH connection to the other nodes of the cluster.
+
+**Note:** This service account can be separate from the service account already used
+for the Mattermost ``systemd`` service itself. It is only important that it is allowed
```
```suggestion
for the Mattermost ``systemd`` service itself. It is important that it is allowed
```

amyblais commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+
+**Note:** This service account can be separate from the service account already used
+for the Mattermost ``systemd`` service itself. It is only important that it is allowed
+to create a SSH tunnel with port forwarding and does not require any additional
```
```suggestion
to create a SSH tunnel with port forwarding, but it does not require any additional
```

amyblais commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+six nodes, use 8075 and 8074 with 1 to 5 in front of it. If the cluster is of bigger size, additional
+ports must be used.
+
+Ensure that your operating system has IP forwarding enabled using the following command:
```
```suggestion
Ensure that your operating system has IP forwarding enabled using the following command:

```

amyblais commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+After that, reload the ufw rules and confirm that the iptable rules were successfully
+created:
+
+.. code-block:: none
```
Code block is not rendering.

amyblais commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+
+
+**Note:** We are ignoring the error from the SSH connection itself in case a tunnel
+is already active. Otherwise the Mattermost server would refuse to start.
```
```suggestion
is already active. Otherwise the Mattermost server would fail to start.
```

amyblais commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+  $ chmod +x /opt/mattermost/bin/pre_start.sh
+
+
+Open the systemd unit file of Mattermost and search for ``Type=Notify``, after it enter
```
```suggestion
Open the systemd unit file of Mattermost and search for ``Type=Notify``. After this, enter
```

amyblais commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+Cluster Start
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Once each node is configured, restart the service on each cluster.
```
```suggestion
Once each node is configured, restart the service on each cluster
```

amyblais commented on source/install/transport-encryption/config-cluster.rst:
```diff
@@ -0,0 +1,231 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Once each node is configured, restart the service on each cluster.
+and confirm it's running using the command below.
```
```suggestion
and confirm that it's running using the command below:
```

amyblais commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,115 @@
+=============================================
+
+Mattermost is able to encrypt the traffic between the database and the application
+using TLS. This guide describes the set up steps for a single, separate MySQL
```
```suggestion
using TLS. This guide describes the setup steps for a single, separate MySQL
```

amyblais commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,115 @@
+using TLS. This guide describes the set up steps for a single, separate MySQL
+server.
+
+## Prerequisites
```
Header formatting should be fixed.

amyblais commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,115 @@
+server uses to encrypt the connection. If you would like to use certificates
+from your company CA, please follow the MySQL documentation for configuration steps.
+
+**Optionally**, it can be enforced that all connections must be made via a local
```
```suggestion
**Note:** Optionally, it can be enforced that all connections must be made via a local
```

amyblais commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,115 @@
+- skip-verify (Require TLS + accept self-signed)
+- preferred (Try TLS, fallback to unencrypted)
+
+In our case we need to use ``skip-verify`` since we use a self-signed certificate,
```
```suggestion
In our case we need to use ``skip-verify`` since we use a self-signed certificate.
```

amyblais commented on source/install/transport-encryption/config-database.rst:
```diff
@@ -0,0 +1,115 @@
+- preferred (Try TLS, fallback to unencrypted)
+
+In our case we need to use ``skip-verify`` since we use a self-signed certificate,
+the configuration setting will now look like this:
```
```suggestion
The configuration setting will now look like this:
```

amyblais commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,102 @@
+Mattermost is able to encrypt the traffic between the proxy and the application
+server using TLS.
+
+## Prerequisites
```
Header formatting should be fixed.

amyblais commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,102 @@
+
+Configuring NGINX
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+On the NGINX server, connect to both servers with a sudo or root user. Open the Mattermost proxy configuration (Refer to - `Configuring NGINX with SSL and HTTP/2 (End-User - Proxy) </install/install-rhel-6.html#configuring-nginx-with-ssl-and-http-2>`__ for additional information)
```
```suggestion
On the NGINX server, connect to both servers with a sudo or root user. Open the Mattermost proxy configuration (Refer to `Configuring NGINX with SSL and HTTP/2 (End-User - Proxy) </install/install-rhel-6.html#configuring-nginx-with-ssl-and-http-2>`__ for additional information)
```

Also, link is not working.

amyblais commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,102 @@
+
+  proxy_pass http://backend;
+
+And change the protocol from ``http`` to ``https``:
```
```suggestion
Also, change the protocol from ``http`` to ``https``:
```

grundleborg commented on source/install/transport-encryption/config.rst:
```diff
@@ -0,0 +1,22 @@
+
+The components of the Mattermost setup are shown in the following diagram,
+including the transport encryption used. Aside from the encryption between the
+Mattermost Application Server, all transports rely on TLS encryption.
```
between the Mattermost Application Server and *what*?

grundleborg replied:
Guessing this should be `Mattermost Application Servers` but I'm certain?

DSchalla replied:
Edited to `Aside from the encryption between the nodes
of the Mattermost cluster, all transports rely on TLS encryption.`, which should make that more clear.

grundleborg commented on source/install/transport-encryption/config.rst:
```diff
@@ -0,0 +1,22 @@
+
+**Note:** The transport between the Application servers is not used by default and
+requires additional setup steps. Enhancing the core product to include automatic
+encryption between cluster nodes is in progress and planned for the 5.18 release.
```
Did this land for 5.18?

DSchalla replied:
Thanks, I suspect it will be 5.20 - Edited!

grundleborg commented on source/install/transport-encryption/config.rst:
```diff
@@ -0,0 +1,22 @@
+
+.. figure:: ../../images/transport-encryption.png
+
+The configuration guides are split up in the following documents:
```
```suggestion
The configuration guides are split up into the following documents:
```

grundleborg commented on source/install/transport-encryption/config-mattermost.rst:
```diff
@@ -0,0 +1,103 @@
+
+  proxy_pass http://backend;
+
+Also, change the protocol from ``http`` to ``https``:
```
```suggestion
Change the protocol from ``http`` to ``https``:
```

sadohert commented:
Hey @DSchalla - Is this close to being published?  If so I'll hold off on sharing with a customer... but if we're still reviewing and discussing I'll share the PR and frame it as draft.

DSchalla commented:
Thanks everyone for the good review! @wiersgallak Could you give it a review from a PM perspective? Otherwise good to merge.
